pkg/storage/iterator.go

```diff
@@ -297,23 +297,44 @@ func newChunksIterator(ctx context.Context, chunks []*chunkenc.LazyChunk, matche
 		}
 	}
 
+	//validChunks := make(map[*chunkenc.LazyChunk])
 	// Finally we load all chunks not already loaded
 	if err := fetchLazyChunks(ctx, allChunks); err != nil {
 		return nil, err
 	}
 
-	iters, err := buildIterators(ctx, chksBySeries, filter, direction, from, through)
+	var validChunks map[*chunkenc.LazyChunk]bool
+	for _, series := range chksBySeries {
+		for _, chunks := range series {
+			for _, chunk := range chunks {
+				if contains(chunk, allChunks) {
+					validChunks[chunk] = true
+				}
+			}
+		}
+	}
+
+	iters, err := buildIterators(ctx, chksBySeries, validChunks, filter, direction, from, through)
```

Comment:
@cyriltovena This is one way we can check and avoid iterating invalid chunks. However, this is going to be O(n^3) operation. 

We can add a label to `LazyChunk` like below but we will have to change many references and disturb the code a bit.

```
type LazyChunk struct {
	Chunk   chunk.Chunk
         IsValid   bool
	Fetcher *chunk.Fetcher
}
```

What do you suggest?

Reply:
```
type LazyChunk struct {
	Chunk   chunk.Chunk
        IsValid   bool
	Fetcher *chunk.Fetcher
}
```

Is less intrusive for this code base IMO I prefer this.

Reply:
Sure. Thanks